### PR TITLE
[hotfix][ST-Engine] Fix user-defined threads cannot use engine metrics

### PR DESCRIPTION
**Pull request by ic4y**
## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

### CODE REVIEW

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/metrics/MetricsContext.java [hunk 1]
```diff
@@ -159,6 +163,65 @@ public void decrement(long decrement) {
...
+
+        private volatile long timestamp;
+
+        ThreadSafeQPSMetric(String name, Unit unit) {
+            super(name, unit);
+        }
+
+        @Override
+        public void increment() {
+            checkAndSetStartTime();
```
Comment (Hisoka-X) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/metrics/MetricsContext.java]: Why not set timestamp in constructor?

**Review by EricJoy2048 — COMMENTED** [review]
Can you add a performance testing for this change?

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/metrics/MetricsContext.java (hunk 1 above)
Comment (ic4y) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/metrics/MetricsContext.java]: Is it more accurate to use the first write time as the start time to calculate QPS?

**Comment by ic4y** [comment]
> Can you add a performance testing for this change?

I will do a performance test today

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/metrics/MetricsContext.java (hunk 1 above)
Comment (ic4y) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/metrics/MetricsContext.java]: You are right, first write time and constructor are close

**Comment by ic4y** [comment]
> > Can you add a performance testing for this change?
> 
> I will do a performance test today

In the file to file synchronization scenario, the amount of data is 1000w, the performance difference between threadSafeMetric and Metric is very small, and the qps are all around 100w/s

```
Metric:
    SinkWriteCount 10000000 
    SinkWriteQPS 1095827.71
    SourceReceivedCount 10000000
    SourceReceivedQPS 1093022.28

ThreadSafeMetric:
    SinkWriteCount 10000000
    SinkWriteQPS 1074834.02
    SourceReceivedCount 10000000     
    SourceReceivedQPS 1071665.76

```

**Comment by EricJoy2048** [comment]
> 1071665.76



> > > Can you add a performance testing for this change?
> > 
> > 
> > I will do a performance test today
> 
> In the file to file synchronization scenario, the amount of data is 100w, the performance difference between threadSafeMetric and Metric is very small, and the qps are all around 100w/s
> 
> ```
> Metric:
>     SinkWriteCount 10000000 
>     SinkWriteQPS 1095827.71
>     SourceReceivedCount 10000000
>     SourceReceivedQPS 1093022.28
> 
> ThreadSafeMetric:
>     SinkWriteCount 10000000
>     SinkWriteQPS 1074834.02
>     SourceReceivedCount 10000000     
>     SourceReceivedQPS 1071665.76
> ```

ThreadSafeMetric 3% slow than Metric. I think it's acceptable. @TyrantLucifer  What do you think?

**Review by EricJoy2048 — APPROVED** [review]

**Review by TyrantLucifer — APPROVED** [review]

**Pull request merged**